lib/core/CDataFrame.cc

```diff
@@ -209,6 +290,33 @@ void CDataFrame::writeRow(const TWriteFunc& writeRow) {
     (*m_Writer)(writeRow);
 }
 
+void CDataFrame::columnNames(TStrVec columnNames) {
+    if (columnNames.size() != m_NumberColumns) {
+        HANDLE_FATAL(<< "Internal error: expected '" << m_NumberColumns << "' column names values but got "
+                     << CContainerPrinter::print(columnNames));
+    } else {
+        m_ColumnNames = std::move(columnNames);
+    }
+}
+
+void CDataFrame::emptyIsMissing(TBoolVec emptyIsMissing) {
+    if (emptyIsMissing.size() != m_NumberColumns) {
+        HANDLE_FATAL(<< "Internal error: expected '" << m_NumberColumns
+                     << "' 'empty is missing' column indicator values but got "
+                     << CContainerPrinter::print(emptyIsMissing));
+    } else {
+        m_EmptyIsMissing = std::move(emptyIsMissing);
+    }
+}
+
+void CDataFrame::categoricalColumns(const TStrVec& categoricalColumnNames) {
```

Comment:
Good observation! In fact, this PR just moved this code out of `api::CDataFrameAnalyzer`. I think this probably would be fine: this is equal to the number of features and only gets called once. That said it would be trivial to sort the vector first, so that change seems prudent.

Reply:
See [5d1a4f0](https://github.com/elastic/ml-cpp/pull/723/commits/5d1a4f0664a89b190c635cc71f0065576424b9dc)